pkg/driver/controller.go

```diff
@@ -815,8 +841,39 @@ func (cs *controller) DeleteSnapshot(
 		// should succeed when an invalid snapshot id is used
 		return &csi.DeleteSnapshotResponse{}, nil
 	}
+	volumeID := snapshotID[0]
 	unlock := cs.volumeLock.LockVolumeWithSnapshot(snapshotID[0], snapshotID[1])
 	defer unlock()
+
+	// verify if the snapshot has already been deleted
+	snap, err := zfs.GetZFSSnapshot(snapshotID[1])
+	if snap != nil && snap.DeletionTimestamp != nil {
+		return csipayload.NewDeleteSnapshotResponseBuilder().Build(), nil
+	}
+
+	if err != nil {
+		if k8serror.IsNotFound(err) {
+			return csipayload.NewDeleteSnapshotResponseBuilder().Build(), nil
+		}
```

Comment:
Isn't there a race condition here, if the pod crashes after issuing `DeleteSnapshot`. Then when it restarts, if it finds the timestamp is set or snapshot is already deleted, it doesn't delete the volume?

Reply:
removed the condition